src/parser/jobs/rpr/modules/HarvestMoon.tsx

```diff
@@ -0,0 +1,84 @@
+import {Trans} from '@lingui/react'
+import {DataLink} from 'components/ui/DbLink'
+import {Event} from 'event'
+import {Analyser} from 'parser/core/Analyser'
+import {filter} from 'parser/core/filter'
+import {dependency} from 'parser/core/Injectable'
+import Checklist, {Requirement, Rule} from 'parser/core/modules/Checklist'
+import {Data} from 'parser/core/modules/Data'
+import {Invulnerability} from 'parser/core/modules/Invulnerability'
+import {UnableToAct} from 'parser/core/modules/UnableToAct'
+import React from 'react'
+
+const SOULSOW_BUFFER = 1000
+
+export class HarvestMoon extends Analyser {
+	static override handle = 'harvestMoon'
+
+	@dependency private checklist!: Checklist
+	@dependency private data!: Data
+	@dependency private invulnerability!: Invulnerability
+	@dependency private unableToAct!: UnableToAct
+
+	private harvestMoonCasts = 0
+
+	override initialise() {
+		super.initialise()
+
+		this.addEventHook(
+			filter<Event>()
+				.source(this.parser.actor.id)
+				.type('action')
+				.action(this.data.actions.HARVEST_MOON.id),
+			() => { this.harvestMoonCasts++ })
+
+		this.addEventHook('complete', this.onComplete)
+	}
+
+	private getExpectedUses(): number {
+		const ADJUSTED_CAST = this.data.actions.SOULSOW.castTime + SOULSOW_BUFFER
+		const invulnWindows = this.invulnerability.getWindows().filter((w) => w.end - w.start >=  ADJUSTED_CAST)
+
+		if (this.unableToAct.getDuration() > 0) {
+			return invulnWindows.filter((invulnWindow) => {
+				const unableToActWindow = this.unableToAct.getWindows({start: invulnWindow.start, end: invulnWindow.end})[0]
+
+				return unableToActWindow == null
+					?false
+					:unableToActWindow.start - invulnWindow.start >= ADJUSTED_CAST
+						||unableToActWindow.end - invulnWindow.end >= ADJUSTED_CAST
```

Comment:
Is there a way to structure this better? Whitespace aside, the logic here is difficult to follow the way it's written.

Reply:
Honestly, I'm not sure. As is, the way I see the problem, there are two cases where you should be using HM during downtime, and a bunch where you shouldn't, so hard icing those two cases (5s before a UTA, or 5s after a UTA.)
That said, I'd be happy to discuss it further in discord, brainstorm ideas to figure out a cleaner way to do it.

Reply:
Regarding readability, I see that I made a minor mistake, whitespace-wise, with following the style of Analyser.ts's ternaries, so I'll fix that.
Do you have any other ideas regarding ways that this logic could be made clearer?

Reply:
I ran into something similar when calculating drift in the Aetherflow module for Scholar. Using Aetherrflow and the core CDDT module as inspiration you could maybe do something like:

```
getExpectedUses(): number {
  let currentTimestamp = this.parser.pull.start
  let expectedUses = 0
  const ADJUSTED_CAST = this.data.actions.SOULSOW.castTime + SOULSOW_BUFFER
  const pullEndTimestamp = this.parser.pull.start + this.parser.pull.duration
  
  while (currentTimestamp < pullEndTimestamp) {
    expectedUses++
  
    currentTimestamp = currentTimestamp + ADJUSTED_CAST
  
    // This is example logic, I'm not super familiar with the edge cases so it's probably more complicated than this
    // The idea would be to adjust current timestamp to wherever the expected cast should be based on downtime edge cases
    if (this.unableToAct.isUnableToAct({ timestamp: currentTimestamp})) {
      const unableToActWindow = this.unableToAct.getWindows({start: currentTimestamp, end: currentTimestamp})[0]
      currentTimestamp = unableToActWindow.end + SOULSOW_BUFFER
    }
  }

  return expectedUses
}
```

Might need a little bit of sprucing up to get it to compile and cover any edge cases I missed.

Reply:
I'll do some testing